> but can you move the declaration of the queue back to `‎system/lib/libc/dynlink.c`?

Unfortunately, this won't work because `em_proxying_queue` is only forward-declared and therefore an incomplete type.
```
../../../system/lib/libc/dynlink.c:362:19: error: variable has incomplete type 'em_proxying_queue' (aka 'struct em_proxying_queue')
  362 | em_proxying_queue _dlopen_proxying_queue = {
      |                   ^
/home/kleisauke/emscripten/cache/sysroot/include/emscripten/proxying.h:29:16: note: forward declaration of 'struct em_proxying_queue'
   29 | typedef struct em_proxying_queue em_proxying_queue;
      |                ^
1 error generated.
```

--- a/system/lib/libc/dynlink.c
+++ b/system/lib/libc/dynlink.c
@@ -356,13 +356,19 @@ static void thread_sync_done(void* arg) {
   free(info);
 }
 
+// Proxying queue specially for handling code loading (dlopen) events.
+// Initialized by the main thread on the first call to
+// `_emscripten_proxy_dlsync` below, and processed by background threads
+// that call `_emscripten_process_dlopen_queue` during futex_wait (i.e. whenever
+// they block).
+em_proxying_queue* _dlopen_proxying_queue = NULL;
 static thread_local bool processing_queue = false;
 
 void _emscripten_process_dlopen_queue() {
-  if (!processing_queue) {
+  if (_dlopen_proxying_queue && !processing_queue) {
     assert(!emscripten_is_main_runtime_thread());
     processing_queue = true;
-    emscripten_proxy_execute_queue(&_dlopen_proxying_queue);
+    emscripten_proxy_execute_queue(_dlopen_proxying_queue);
     processing_queue = false;
   }
 }
@@ -373,6 +379,9 @@ void _emscripten_process_dlopen_queue() {
 // manages the worker pool.
 int _emscripten_proxy_dlsync_async(pthread_t target_thread, em_promise_t promise) {
   assert(emscripten_is_main_runtime_thread());
+  if (!_dlopen_proxying_queue) {
+    _dlopen_proxying_queue = em_proxying_queue_create();
+  }
 
   struct promise_result* info = malloc(sizeof(struct promise_result));
   if (!info) {
@@ -382,7 +391,7 @@ int _emscripten_proxy_dlsync_async(pthread_t target_thread, em_promise_t promise
     .promise = promise,
     .result = false,
   };
-  int rtn = emscripten_proxy_callback(&_dlopen_proxying_queue,
+  int rtn = emscripten_proxy_callback(_dlopen_proxying_queue,
                                       target_thread,
                                       do_thread_sync,
                                       thread_sync_done,
@@ -400,9 +409,12 @@ int _emscripten_proxy_dlsync_async(pthread_t target_thread, em_promise_t promise
 
 int _emscripten_proxy_dlsync(pthread_t target_thread) {
   assert(emscripten_is_main_runtime_thread());
+  if (!_dlopen_proxying_queue) {
+    _dlopen_proxying_queue = em_proxying_queue_create();
+  }
   int result;
   if (!emscripten_proxy_sync(
-        &_dlopen_proxying_queue, target_thread, do_thread_sync_out, &result)) {
+        _dlopen_proxying_queue, target_thread, do_thread_sync_out, &result)) {
     return 0;
   }
   return result;

--- a/system/lib/libc/dynlink.c
+++ b/system/lib/libc/dynlink.c
@@ -361,7 +361,7 @@ static void thread_sync_done(void* arg) {
 // `_emscripten_proxy_dlsync` below, and processed by background threads
 // that call `_emscripten_process_dlopen_queue` during futex_wait (i.e. whenever
 // they block).
-em_proxying_queue* _dlopen_proxying_queue = NULL;
+em_proxying_queue* _Atomic _dlopen_proxying_queue = NULL;
 static thread_local bool processing_queue = false;
 
 void _emscripten_process_dlopen_queue() {

--- a/system/lib/pthread/proxying.c
+++ b/system/lib/pthread/proxying.c
@@ -41,18 +41,6 @@ em_proxying_queue* emscripten_proxy_get_system_queue(void) {
   return &system_proxying_queue;
 }
 
-#ifdef EMSCRIPTEN_DYNAMIC_LINKING
-// Proxying queue specially for handling code loading (dlopen) events.
-// Processed by background threads that call `_emscripten_process_dlopen_queue`
-// during futex_wait (i.e. whenever they block).
-em_proxying_queue _dlopen_proxying_queue = {
-  .mutex = PTHREAD_MUTEX_INITIALIZER,
-  .task_queues = NULL,
-  .size = 0,
-  .capacity = 0,
-};
-#endif
-
 em_proxying_queue* em_proxying_queue_create(void) {
   // Allocate the new queue.
   em_proxying_queue* q = malloc(sizeof(em_proxying_queue));
@@ -182,7 +170,7 @@ static bool do_proxy(em_proxying_queue* q, pthread_t target_thread, task t) {
   // requires a wakeup of the target thread after enqueuing work.
   bool needs_notify =
 #ifdef EMSCRIPTEN_DYNAMIC_LINKING
-    q == &_dlopen_proxying_queue ||
+    q == _dlopen_proxying_queue ||
 #endif
     (is_system_queue &&
      pthread_equal(target_thread, emscripten_main_runtime_thread_id()));

--- a/system/lib/pthread/threading_internal.h
+++ b/system/lib/pthread/threading_internal.h
@@ -86,7 +86,7 @@ void _emscripten_thread_exit(void* result);
 
 #ifdef EMSCRIPTEN_DYNAMIC_LINKING
 void _emscripten_process_dlopen_queue(void);
-extern em_proxying_queue _dlopen_proxying_queue;
+extern em_proxying_queue* _dlopen_proxying_queue;
 #endif
 
 #if !defined(__EMSCRIPTEN_PTHREADS__) || defined(NDEBUG)

--- a/system/lib/pthread/threading_internal.h
+++ b/system/lib/pthread/threading_internal.h
@@ -83,11 +83,8 @@ int _emscripten_thread_is_valid(pthread_t thread);
 
 void _emscripten_thread_exit_joinable(pthread_t thread);
 void _emscripten_thread_exit(void* result);
-
-#ifdef EMSCRIPTEN_DYNAMIC_LINKING
 void _emscripten_process_dlopen_queue(void);
 extern em_proxying_queue* _Atomic _dlopen_proxying_queue;
-#endif
 
 #if !defined(__EMSCRIPTEN_PTHREADS__) || defined(NDEBUG)
 #define emscripten_set_current_thread_status(newStatus)

--- a/system/lib/pthread/threading_internal.h
+++ b/system/lib/pthread/threading_internal.h
@@ -86,7 +86,7 @@ void _emscripten_thread_exit(void* result);
 
 #ifdef EMSCRIPTEN_DYNAMIC_LINKING
 void _emscripten_process_dlopen_queue(void);
-extern em_proxying_queue* _dlopen_proxying_queue;
+extern em_proxying_queue* _Atomic _dlopen_proxying_queue;
 #endif
 
 #if !defined(__EMSCRIPTEN_PTHREADS__) || defined(NDEBUG)

